test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -34,10 +34,19 @@ final case class Spec[-R, +E, +L, +T](caseValue: SpecCase[R, E, L, T, Spec[R, E,
    * test("foo") { assert(42, equalTo(42)) } @@ ignore
    * }}}
    */
-  final def @@[R0 <: R1, R1 <: R, E0 >: E, E1 >: E0, S0 <: S, S, S1 >: S](aspect: TestAspect[R0, R1, E0, E1, S0, S1])(
-    implicit ev: T <:< Either[TestFailure[Nothing], TestSuccess[S]]
-  ): Spec[R1, E0, L, Either[TestFailure[Nothing], TestSuccess[S]]] =
-    aspect(mapTest(ev))
+  final def @@[R0 <: R1, R1 <: R, E0, E1, E2 >: E0 <: E1, S0, S1, S >: S0 <: S1](
+    aspect: TestAspect[R0, R1, E0, E1, S0, S1]
+  )(implicit ev1: E <:< TestFailure[E2], ev2: T <:< TestSuccess[S]): ZSpec[R1, E2, L, S] =
```

Comment:
While generally I like implicit evidence, given the cost of the bimap, may be worth pushing `@@` to be a syntax (extension method) in the companion object, which is added only to specs with `TestFailure` and `TestSuccess` as the S / T. Alternately we could consider specializing all of `Spec` to these types, but the more polymorphic version is nicer because most methods don't need to know about that structure.

Reply:
If we are worried about performance could we use `asInstanceOf` since we have the implicit evidence?

I explored specialized `Spec` when I was working on this PR and I don't think we want to go down that road. If we specialize `Spec` on a type that has a error we lose the ability to set the error channel to `Nothing` in `ExecutedSpec` to express that those specs can't fail, which is quite important now that they are effectual to be able to safely traverse them.

Also we moved `@@` to `Spec` originally to solve type inference issues, so not to say we can't come up with another way but I'm worried we are going to take a step back there.

Reply:
All right, cast sounds good to me as long as it is sound.